libs/vkd3d/command.c

```diff
@@ -11369,58 +11414,47 @@ static void STDMETHODCALLTYPE d3d12_command_list_WriteBufferImmediate(d3d12_comm
-            if (flush_before)
-            {
-                if (dword_count)
-                {
-                    d3d12_command_list_mark_copy_buffer_write(list, curr_buffer,
-                            curr_offset, dword_count * sizeof(uint32_t), false);
-
-                    VK_CALL(vkCmdUpdateBuffer(list->vk_command_buffer, curr_buffer,
-                            curr_offset, dword_count * sizeof(uint32_t), dword_buffer));
-                }
-
-                curr_buffer = resource->vk_buffer;
-                curr_offset = offset;
-                dword_count = 0;
-            }
-
-            dword_buffer[dword_count++] = parameters[i].Value;
-
-            /* Record batched writes if the buffer is full or if we're at the
-             * end of the list, or if the next write has marker semantics. */
-            flush_after = dword_count == ARRAY_SIZE(dword_buffer) || i + 1 == count ||
-                    (modes && modes[i + 1] != mode && list->device->vk_info.AMD_buffer_marker);
+        /* Avoid ending the active render pass instance, if any */
+        do_batch = mode == D3D12_WRITEBUFFERIMMEDIATE_MODE_DEFAULT
+                || !list->device->vk_info.AMD_buffer_marker
+                || (list->wbi_batch.batch_len && mode != D3D12_WRITEBUFFERIMMEDIATE_MODE_MARKER_IN);
 
-            if (flush_after)
-            {
-                d3d12_command_list_mark_copy_buffer_write(list, curr_buffer,
-                        curr_offset, dword_count * sizeof(uint32_t), false);
+        if (do_batch)
+        {
+            batch_entry = list->wbi_batch.batch_len++;
 
-                VK_CALL(vkCmdUpdateBuffer(list->vk_command_buffer, curr_buffer,
-                        curr_offset, dword_count * sizeof(uint32_t), dword_buffer));
+            list->wbi_batch.buffers[batch_entry] = resource->vk_buffer;
+            list->wbi_batch.offsets[batch_entry] = offset;
+            list->wbi_batch.stages[batch_entry] = stage;
+            list->wbi_batch.values[batch_entry] = parameters[i].Value;
 
-                curr_buffer = VK_NULL_HANDLE;
-                curr_offset = 0;
-                dword_count = 0;
-            }
+            if (mode == D3D12_WRITEBUFFERIMMEDIATE_MODE_MARKER_IN)
+                do_flush = true;
         }
         else
         {
             VK_CALL(vkCmdWriteBufferMarkerAMD(list->vk_command_buffer, stage,
                     resource->vk_buffer, offset, parameters[i].Value));
         }
+
+        if ((do_flush && i + 1 == count) || list->wbi_batch.batch_len == VKD3D_MAX_WBI_BATCH_SIZE)
+        {
+            if (list->rendering_info.state_flags & VKD3D_RENDERING_ACTIVE)
+            {
+                /* Implicitly calls end_wbi_batch. We cannot have any pending transfers
+                 * while inside a render pass instance. */
+                d3d12_command_list_end_current_render_pass(list, true);
+
+                /* Relevant if we exceeded the maximum batch size */
+                do_flush = true;
```

Comment:
No, the idea is that we will only defer WBI batches if we're inside a render pass instance, but since we're ending the render pass instance here we have to flush any subsequent batch from the current WBI command as well.

Reply:
Ok, right ... That interaction is pretty subtle. Worth a comment I think.

Reply:
I guess the comment was changed from when I last reviewed, now it makes more sense.